File: izmk-common/src/main/java/ovo/xsvf/logging/Logger.java
package ovo.xsvf.logging;

import com.google.gson.JsonObject;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;

import java.io.IOException;
import java.io.OutputStreamWriter;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class Logger {
    private final OutputStreamWriter writer;
    private final String name;

    public static Logger of(String name, int port) {
        Socket socket;
        try {
            Logger logger = new Logger(new OutputStreamWriter(new Socket("localhost", port).getOutputStream(), StandardCharsets.UTF_8), name);
            JsonObject jsonObject = new JsonObject();
            jsonObject.addProperty("type", "init");
            jsonObject.addProperty("name", name);
            logger.sendJson(jsonObject);
            return logger;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static String format(String format, Object... args) {
        for (Object arg : args) {
            format = format.replaceFirst("\\{}", arg.toString());
        }
        return format;
    }

    public void sendJson(JsonObject jsonObject) {
        try {
            writer.write(Base64.getEncoder().encodeToString(jsonObject.toString().getBytes(StandardCharsets.UTF_8)) + "\n");
            writer.flush();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public void log(int level, String message) {
        JsonObject jsonObject = new JsonObject();
        jsonObject.addProperty("type", "log");
        jsonObject.addProperty("level", level);
        jsonObject.addProperty("message", message);

        sendJson(jsonObject);
    }

    public void debug(String message, Object... args) {
        log(Level.DEBUG, format(message, args));
    }

    public void info(String message, Object... args) {
        log(Level.INFO, format(message, args));
    }

    public void warn(String message, Object... args) {
        log(Level.WARN, format(message, args));
    }

    public void error(String message, Object... args) {
        log(Level.ERROR, format(message, args));
    }

    public void error(String msg, Throwable throwable, Object... args) {
        StringBuilder errorMessage = new StringBuilder();
        errorMessage.append(String.format(msg, args))
                .append(System.lineSeparator());
        errorMessage.append("    ").append("Exception: ")
                .append(throwable.getClass().getSimpleName()).append(": ")
                .append(throwable.getMessage() == null ? "" : throwable.getMessage())
                .append(System.lineSeparator());
        Throwable cause = throwable.getCause();
        if (cause != null) {
            errorMessage.append("    ").append("Cause: ")
                    .append(cause.getMessage() == null ? cause.getClass().getSimpleName() : cause.getMessage())
                    .append(System.lineSeparator());
        }
        errorMessage.append("    ").append("Stack Trace: ")
                .append(System.lineSeparator());
        for (StackTraceElement element : throwable.getStackTrace()) {
            errorMessage.append("    ").append("    ").append(element.toString()).append(System.lineSeparator());
        }

        error(errorMessage.toString());
    }

    public void error(Throwable throwable, Object... args) {
        StringBuilder errorMessage = new StringBuilder();
        errorMessage.append("Exception: ")
                .append(throwable.getClass().getSimpleName()).append(": ")
                .append(throwable.getMessage() == null ? "" : throwable.getMessage())
                .append(System.lineSeparator());
        Throwable cause = throwable.getCause();
        if (cause != null) {
            errorMessage.append("    ").append("Cause: ")
                    .append(cause.getMessage() == null ? cause.getClass().getSimpleName() : cause.getMessage())
                    .append(System.lineSeparator());
        }
        errorMessage.append("    ").append("Stack Trace: ")
                .append(System.lineSeparator());
        for (StackTraceElement element : throwable.getStackTrace()) {
            errorMessage.append("    ").append("    ").append(element.toString()).append(System.lineSeparator());
        }

        error(errorMessage.toString());
    }


    public static class Level {
        public static final int DEBUG = 0;
        public static final int INFO = 1;
        public static final int WARN = 2;
        public static final int ERROR = 3;
    }
}

File: izmk-common/src/main/kotlin/ovo/xsvf/logging/Logger.kt
package ovo.xsvf.logging

import com.google.gson.JsonObject
import java.io.BufferedWriter
import java.io.OutputStreamWriter
import java.net.Socket
import java.nio.charset.StandardCharsets
import java.util.*

/**
 * @author xsvf, LangYa466
 */
class Logger private constructor(private val writer: BufferedWriter, private val socket: Socket, private val name: String) {
    companion object {
        fun of(name: String, port: Int): Logger {
            val socket = Socket("localhost", port)
            val writer = BufferedWriter(OutputStreamWriter(socket.getOutputStream(), StandardCharsets.UTF_8))
            val logger = Logger(writer, socket, name)

            logger.sendJson(
                JsonObject().apply {
                    addProperty("type", "init")
                    addProperty("name", name)
                }
            )

            return logger
        }
    }

    private fun sendJson(jsonObject: JsonObject) {
        val base64 = Base64.getEncoder().encodeToString(jsonObject.toString().toByteArray(StandardCharsets.UTF_8))
        writer.write(base64)
        writer.newLine() // 确保每条消息完整
        writer.flush()
    }

    private fun log(level: Int, message: String) {
        sendJson(
            JsonObject().apply {
                addProperty("type", "log")
                addProperty("level", level)
                addProperty("message", message)
            }
        )
    }

    fun debug(message: String, vararg args: Any) = log(Level.DEBUG, message.format2(*args))

    fun info(message: String, vararg args: Any) = log(Level.INFO, message.format2(*args))

    fun warn(message: String, vararg args: Any) = log(Level.WARN, message.format2(*args))

    fun error(message: String, vararg args: Any) = log(Level.ERROR, message.format2(*args))

    fun error(msg: String, throwable: Throwable, vararg args: Any) {
        error(buildErrorMessage(msg.format2(*args), throwable))
    }

    fun error(throwable: Throwable) {
        error(buildErrorMessage("Exception Occurred", throwable))
    }

    private fun buildErrorMessage(msg: String, throwable: Throwable): String {
        return buildString {
            appendLine(msg)
            append("    Exception: ${throwable::class.simpleName}")
            throwable.message?.let { append(": $it") }
            appendLine()
            throwable.cause?.let { appendLine("    Cause: ${it.message ?: it::class.simpleName}") }
            appendLine("    Stack Trace:")
            append(throwable.stackTraceToString().prependIndent("    "))
        }
    }

    private fun String.format2(vararg args: Any): String {
        var index = 0
        return this.replace(Regex("\\{}")) {
            if (index < args.size) args[index++].toString() else "{}"
        }
    }

    fun close() {
        writer.close()
        socket.close()
    }

    object Level {
        const val DEBUG = 0
        const val INFO = 1
        const val WARN = 2
        const val ERROR = 3
    }
}
